lib/SILGen/SILGenPattern.cpp

```diff
@@ -1031,18 +1031,27 @@ void PatternMatchEmission::emitWildcardDispatch(ClauseMatrix &clauses,
   bool hasGuard = guardExpr != nullptr;
   assert(!hasGuard || !clauses[row].isIrrefutable());
 
-  auto stmt = clauses[row].getClientData<CaseStmt>();
-  ArrayRef<CaseLabelItem> labelItems = stmt->getCaseLabelItems();
-  bool hasMultipleItems = labelItems.size() > 1;
-  
+  auto stmt = clauses[row].getClientData<Stmt>();
+  assert(isa<CaseStmt>(stmt) || isa<CatchStmt>(stmt));
+
+  bool hasMultipleItems = false;
+  if (auto *caseStmt = dyn_cast<CaseStmt>(stmt)) {
+    hasMultipleItems = caseStmt->getCaseLabelItems().size() > 1;
+  }
+
   // Bind the rest of the patterns.
   bindIrrefutablePatterns(clauses[row], args, !hasGuard, hasMultipleItems);
 
   // Emit the guard branch, if it exists.
   if (guardExpr) {
-    SGF.usingImplicitVariablesForPattern(clauses[row].getCasePattern(), stmt, [&]{
-      this->emitGuardBranch(guardExpr, guardExpr, failure);
-    });
+    if (auto *caseStmt = dyn_cast<CaseStmt>(stmt)) {
+      SGF.usingImplicitVariablesForPattern(clauses[row].getCasePattern(), caseStmt, [&]{
+        this->emitGuardBranch(guardExpr, guardExpr, failure);
+      });
+    } else {
+      // For CatchStmt
+      emitGuardBranch(guardExpr, guardExpr, failure);
```

Comment:
I'm still wondering why we need to branch—this is probably a hazard elsewhere in code that handles patterns too. I think it would be much more robust for `SGF.usingImplicitVariablesForPattern` to handle the `caseStmt == nullptr` case inside itself instead of delegating this responsibility to every call site, so this becomes:

```swift
      SGF.usingImplicitVariablesForPattern(clauses[row].getCasePattern(), dyn_cast<CaseStmt>(stmt), [&]{
        this->emitGuardBranch(guardExpr, guardExpr, failure);
      });
```

and `SGF.usingImplicitVariablesForPattern` has an early exit like:

```swift
  if (caseStmt == nullptr)
    // Call through to the closure directly.
    return f();
```